rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+
+The `preboot` (see [unresolved question 1](#81-lifecycle-stage-name)) is the Kibana initial lifecycle stage at which it only initializes a bare minimum of the core services and a limited set of special-purpose plugins. It's assumed that Kibana can change and reload its own configuration at this stage and may require administrator involvement before it can proceed to the `setup` and `start` stages.
+
+# 2. Motivation
+
+The `preboot` lifecycle stage is a prerequisite for the Kibana interactive setup mode. This is the mode Kibana enters to on the first launch if it detects that user hasn't explicitly configured their own connection to Elasticsearch. In this mode, Kibana will present an interface to the user that would allow them to provide Elasticsearch connection information and potentially any other configuration information. Once the information is verified, Kibana will write it to the disk and allow the rest of Kibana to start.
+
+The interactive setup mode will be provided through a dedicated `Setup` plugin that will be initialized at the `preboot` stage.
+
+# 3. Detailed design
+
+The central part of the `preboot` stage is a dedicated HTTP server instance formerly known as `Not Ready` server. Kibana starts this server at the `preboot` stage and shuts it down as soon as the main HTTP server is ready to start. Currently, this server only exposes a status endpoint and renders a static `Kibana server is not ready yet` string whenever users try to access Kibana before it's completely initialized.
+
+The changes proposed in this RFC should allow special-purpose plugins to define custom HTTP endpoints, and serve interactive client-side applications on the aforementioned HTTP server, and hence make Kibana interactive setup mode possible.
+
+## 3.1 Core client-side changes
+
+The RFC aims to limit the changes to only those that are absolutely required and doesn't assume any modifications in the client-side part of the Kibana Core at the moment. This may introduce a certain level of inconsistency in the client-side codebase, but we consider it insignificant. See [unresolved question 2](#82-core-client-side-changes) for more details.
+
+## 3.2 Core server-side changes
+
+We'll update only several Core server-side services to support the new `preboot` lifecycle stage and preboot plugins.
+
+Once the `preboot` stage is complete, Kibana might need to reload the configuration before it can proceed to `setup`. This doesn't require any special care from the existing plugin developers since Kibana would instantiate plugins only after it reloads the config. We'll also make sure that neither of the Core services relies on the stale configuration it may have acquired during the `preboot` stage.
+
+### 3.2.1 Plugins service
+
+First of all, we'll introduce a new type of special-purpose plugins: preboot plugins, in contrast to ordinary or primary plugins. Kibana will initialize preboot plugins at the `preboot` stage, before even instantiating primary plugins.
```

Comment:
Yes, I agree. I intentionally omitted plugin-specific configuration in the RFC since we don't have a good solution for this yet. The required core schemas can be provided by the core at the `preboot` stage in one way or another, and when/if we decide to change plugin specific configuration we'll figure this out.